core/client/fs/src/main/java/alluxio/client/file/PositionReadFileInStream.java

```diff
@@ -56,18 +192,76 @@ public int read(byte[] b, int off, int len) throws IOException {
 
   @Override
   public int read(ByteBuffer byteBuffer, int off, int len) throws IOException {
-    int bytesRead = mPositionReader.read(mPos, byteBuffer, len);
-    if (bytesRead <= 0) {
-      return bytesRead;
+    byteBuffer.position(off).limit(off + len);
+    mCache.addTrace(mPos, len);
+    int totalBytesRead = 0;
+    int bytesReadFromCache = mCache.fillWithCache(mPos, byteBuffer);
+    totalBytesRead += bytesReadFromCache;
+    mPos += bytesReadFromCache;
+    if (!byteBuffer.hasRemaining()) {
+      return totalBytesRead;
+    }
+    int bytesPrefetched = mCache.prefetch(mPositionReader, mPos, byteBuffer.remaining());
+    if (bytesPrefetched < 0) {
+      if (totalBytesRead == 0) {
+        return -1;
+      }
+      return totalBytesRead;
+    }
+    bytesReadFromCache = mCache.fillWithCache(mPos, byteBuffer);
+    totalBytesRead += bytesReadFromCache;
+    mPos += bytesReadFromCache;
+    if (!byteBuffer.hasRemaining()) {
+      return totalBytesRead;
+    }
+    int bytesRead = mPositionReader.read(mPos, byteBuffer, byteBuffer.remaining());
```

Comment:
todo: this is reading from the reader twice. The correct logic should be
1. if the caller wants more than the prefetch size, read from the reader with the size of the output buffer directly, and optionally copy into the internal buffer
2. otherwise, prefetch into the internal buffer first, and then copy into the output buffer. if somehow the prefetch acquires less data than the caller wants, then be it and do not start another read from the reader.